src/rtc_pl031.rs

```diff
@@ -0,0 +1,586 @@
+/// # use std::io::{Error, Result};
+/// # use std::ops::Deref;
+/// # use std::time::{Instant, Duration, SystemTime, UNIX_EPOCH};
+/// # use vm_superio::RTC;
+///
+/// let mut data = [0; 4];
+/// let mut rtc = RTC::new();
+/// const RTCDR: u32 = 0x0; // Data Register.
+/// const RTCLR: u32 = 0x8; // Load Register.
+///
+/// // Write system time since UNIX_EPOCH in seconds to the load register.
+/// let v = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
+/// data = (v as u32).to_le_bytes();
+/// rtc.write(RTCLR, &data).unwrap();
+///
+/// // Read the value back out of the load register.
+/// rtc.read(RTCLR, &mut data);
+/// assert_eq!((v as u32), u32::from_le_bytes(data));
+///
+/// // Sleep for 1.5 seconds to let the counter tick.
+/// let delay = Duration::from_millis(1500);
+/// thread::sleep(delay);
+///
+/// // Read the current RTC value from the Data Register
+/// rtc.read(RTCDR, &mut data);
+/// assert!(u32::from_le_bytes(data) > (v as u32));
+/// ```
+pub struct RTC {
+    // Counts up from 1 on reset at 1Hz (emulated).
+    counter: Instant,
+
+    // The offset value applied to the counter to get the RTC value.
+    lr: u32,
+
+    // The MR register is used for implementing the RTC alarm. A
+    // real time clock alarm is a feature that can be used to allow
+    // a computer to 'wake up' after shut down to execute tasks
+    // every day or on a certain day. It can sometimes be found in
+    // the 'Power Management' section of a motherboard's BIOS setup.
+    // This is not currently implemented, so we raise an error.
+    // TODO: Implement the match register functionality.
+    mr: u32,
+
+    // The interrupt mask.
+    imsc: u32,
+
+    // The raw interrupt value.
+    ris: u32,
+}
+
+/// Errors encountered while handling RTC operations.
+#[derive(Debug)]
+pub enum Error {
+    /// Attempting to write to a non-existent register offset
+    /// or to the match register, which is not emulated.
+    BadWriteOffset(u32),
+}
+
+impl RTC {
+    /// Creates a new `AMBA PL031 RTC` instance.
+    ///
+    /// # Example
+    ///
+    /// You can see an example of how to use this function in the
+    /// [`Example` section from `RTC`](struct.RTC.html#example).
+    pub fn new() -> RTC {
+        RTC {
+            // Counts up from 1 on reset at 1Hz (emulated).
+            counter: Instant::now(),
+
+            // The load register is initialized to zero.
+            lr: 0,
+
+            // The match register is initialised to zero (not currently used).
+            // TODO: Implement the match register functionality.
+            mr: 0,
+
+            // The interrupt mask is initialised as not set.
+            imsc: 0,
+
+            // The raw interrupt is initialised as not asserted.
+            ris: 0,
+        }
+    }
+
+    fn get_rtc_value(&self) -> u32 {
+        // Add the counter offset to the seconds elapsed since reset.
+        (self.counter.elapsed().as_secs() as u32) + self.lr
```

Comment:
There is a (very) low possibility of overflow here. In debug builds, this would result in a panic. In release builds the expected behavior is to wrap around. To make sure that this happens, and avoid any possible misuse, can we use explicit wrap around operations for this add?

Reply:
I've changed this to be explicit.  It now uses `wrapping_add`:
```
(self.counter.elapsed().as_secs() as u32).wrapping_add(self.lr)
```

I've also added a test called `test_rtc_value_overflow()` which loads `u32::MAX` into the load register, reads the data register to make sure it's properly loaded, waits 1.5 seconds, and then reads the data register again to confirm it's wrapped around.